src/pd.rs

```diff
@@ -0,0 +1,226 @@
+use crate::pd::pd3p2::SprProgrammableSupplyPdo as Pd3p2SprProgrammableSupplyPdo;
+use crate::pd::pd3p2::VariableSupplyPdo as Pd3p2VariableSupplyPdo;
+
+#[cfg(feature = "c_api")]
+mod c_api {
+    pub(crate) use crate::pd::pd3p2::Pd3p2BatteryCapData;
+    pub(crate) use crate::pd::pd3p2::Pd3p2BatteryStatusData;
+    pub(crate) use crate::pd::pd3p2::Pd3p2BatterySupplyPdo;
+    pub(crate) use crate::pd::pd3p2::Pd3p2DiscoverIdentityResponse;
+    pub(crate) use crate::pd::pd3p2::Pd3p2FixedSupplyPdo;
+    pub(crate) use crate::pd::pd3p2::Pd3p2RevisionMessageData;
+    pub(crate) use crate::pd::pd3p2::Pd3p2SinkCapabilitiesExtended;
+    pub(crate) use crate::pd::pd3p2::Pd3p2SourceCapabilitiesExtended;
+    pub(crate) use crate::pd::pd3p2::Pd3p2SprProgrammableSupplyPdo;
+    pub(crate) use crate::pd::pd3p2::Pd3p2VariableSupplyPdo;
+}
+
+pub mod pd3p2;
+
+#[derive(Debug, Clone, PartialEq, Default, N, CApiWrapper)]
+#[c_api(prefix = "Pd", repr_c)]
+pub enum CommandType {
+    /// Request from initiator port.
+    #[default]
+    Request,
+    /// Acknowledge response from responder port.
+    Ack,
+    /// Negative acknowledge response from responder port.
+    Nak,
+    /// Busy response from responder port.
+    Busy,
+}
+
+#[derive(Debug, Clone, PartialEq, Default, N, Copy, CApiWrapper)]
+#[c_api(prefix = "Pd", repr_c)]
+pub enum Command {
+    /// The Discover Identity Command is provided to enable an Initiator to
+    /// identify its Port Partner and for an Initiator (VCONN Source) to
+    /// identify the Responder (Cable Plug or VPD). The Discover Identity
+    /// Command is also used to determine whether a Cable Plug or VPD is
+    /// PD-Capable by looking for a GoodCRC Message Response.
+    #[default]
+    DiscoverIdentity,
+    DiscoverSVIDs,
+    DiscoverModes,
+    EnterMode,
+    ExitMode,
+    Attention,
+    SVIDSpecific,
+}
+
+#[derive(Debug, Clone, PartialEq, Default, CApiWrapper)]
+#[c_api(prefix = "Pd", repr_c)]
+/// The VDM header. See table 6.30 in the USB PD Specification for more
+/// information.
+pub struct VdmHeader {
+    // Whether this is a structured VDM.
+    pub structured: bool,
+    // The major version number of this VDM.
+    pub major: u8,
+    // Them minor major version number of this VDM.
+    pub minor: u8,
+    /// For Enter Mode, Exit Mode and Attention commands:
+    ///
+    /// Index into the list of VDOs to identify the desired Mode
+    ///
+    /// For Exit Mode only: 0b111 to exit all Active Modes
+    ///
+    /// Zero otherwise.
+    pub object_position: u8,
+    /// The command type.
+    pub command_type: CommandType,
+    /// The command.
+    pub command: Command,
+}
+
+#[derive(Debug, Clone, PartialEq, CApiWrapper)]
+#[c_api(prefix = "Pd", repr_c)]
+pub enum Pdo {
+    Pd3p2FixedSupplyPdo(Pd3p2FixedSupplyPdo),
+    Pd3p2BatterySupplyPdo(Pd3p2BatterySupplyPdo),
+    Pd3p2VariableSupplyPdo(Pd3p2VariableSupplyPdo),
+    Pd3p2AugmentedPdo(Pd3p2SprProgrammableSupplyPdo),
+}
```

Comment:
I wonder about the integration plans for other PD revisions. Would it be sensible to move this into the pd3p2 file and then have a Pdo enum, which contains all of the sub-enums? I.e.

In pd3p2 file:
```
pub enum Pdo {
    FixedSupplyPdo(FixedSupplyPdo),
    BatterySupplyPdo(BatterySupplyPdo),
    VariableSupplyPdo(VariableSupplyPdo),
    SprProgrammableSupplyPdo(SprProgrammableSupplyPdo),
}
```

here:
```
pub enum Pdo {
    Pd3p2(pd3p2::Pdo),
    Pd2p0(pd2p0::Pdo),
}
```

Reply:
Well, my original idea would be

```
Pd3p0FixedSupplyPdo(Pd3p0FixedSupplyPdo)
... so on
```

I agree, what you proposed is cleaner. This requires yet more changes in the CApiWrapper proc-macro though, which is a bit finicky. I can provide that as a follow-up patch since the current code will work as-is even with more versions added.

I want to focus on getting this merged upstream first, since we're a little behind in that aspect.